Tooltip smart position

venetak opened the pull request:
- Change the position of the tooltip if it isn't visible with the current settings.
- Use only the pre-defined positions as they are all anchored on the tooltip target, If we move the tooltip too much it  might become unclear which is its target, for example in a form with multiple elements which are close to each other.
- Check if the tooltip overflows out of the windows, not its container element, as all tooltips are absolutely positioned to the body. 
- In case of need of repositioning we try all possible orientations (top, left, right and bottom) until wee either find  one that is suitable or we've tried all and none fixed the issue.
- Added test case for one position as the logic is the same for all and there's no need to check them individually.

venetak commented on components/tooltip/script.js:
```diff
@@ -61,12 +63,24 @@ class Tooltip extends HTMLElement {
         this.visible = false;
+        // reset the current positioning as the page might get
+        // resized untill the next time the tooltip is displayed
+        this.fixedSides = [];
```
There are the orientations (top, left, bottom, right) that we've tried already. We check all possible positions subsequently until we either find one that fixes the hidden tooltip or we run out of options. We need to track which orientations we've tries so far so that we don't try the same position multiple times or run over all them indefinitely.

The name is unclear and confusing, would appreciate suggestions.

MihailTodorov replied:
checkedOrientations ?

venetak replied:
Yes! A lot better, thanks.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -75,7 +89,10 @@ class Tooltip extends HTMLElement {
             left: elementSize.left + (elementSize.width / 2) - tooltipSize.width / 2
         }
```
Can't we rename this variable as elementPosition, as it is very similar to this.position?

venetak replied:
Yes - changed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+        this.style.left = scrollOffsetX + position.left + 'px';
+
+        // check if the tooltip still overflows and re-position it again
+        const overflows  = this.overflows();
```
```suggestion
        const overflows = this.overflows();

```

venetak replied:
Fixed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+
+        // check if the tooltip still overflows and re-position it again
+        const overflows  = this.overflows();
+        if(Object.keys(overflows).length && this.fixedSides.length !== 4) {
```
Can you set this as a variable for easier reading?

venetak replied:
Yes - assigned `Object.keys(overflows)` to a variable in order to keep the .length accessor for consistency with the other check in the expression:

`overflowingSides.length && this.checkedOrientations.length`

Can also change it to:

`overflowingSidesLength && this.checkedOrientations.length`

but it is inconsistent, though I don't have any strong preference.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+        }
+    }
+
+    overflows() {
```
We can change this to a getter to avoid `const overflows = this.overflows()`, but use `this.overflows` directly

venetak replied:
Changed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+    }
+
+    getVisiblePosition(overflows) {
+        const allSides = ['top', 'bottom', 'left', 'right'];
```
We can set this as a global variable

venetak replied:
Moved to a global and renamed to TOOLTIP_POSITIONS, though I wonder if the prefix is redundant since we're already in the context of the tooltip - what do you think? Should I rename TOOLTIP_MARGIN and TOOLTIP_POSITIONS to MARGIN and POSITIONS?

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+        return overflows;
+    }
+
+    getVisiblePosition(overflows) {
```
```suggestion
    getVisiblePosition(overflows = this.overflows()) {

```

venetak replied:
Changed using the getter.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+    getVisiblePosition(overflows) {
+        const allSides = ['top', 'bottom', 'left', 'right'];
+
+        overflows = overflows || this.overflows();
```
We can remove this

venetak replied:
Removed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+
+        overflowingSides = overflowingSides.filter(side => this.fixedSides.indexOf(side) == -1);
+
+        let nextIndex = allSides.indexOf(this.position) + 1;
```
Maybe you wanted to do `overflowingSides` instead of `allSides`?

venetak replied:
Actually, here we don't need overflowingSides at all. We check if the tooltip overflows and then we try one of the other possible positions. I updated the method a little to only try the positions that are not currently overflowing or that haven't been tried yet. This reduced the amount of tries.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -75,7 +89,10 @@ class Tooltip extends HTMLElement {
 
-        switch (this.position) {
+        const scrollOffsetX = window.scrollX;
+        const scrollOffsetY = window.scrollY;
```
We can move these both to the show() method to set them once on show instead on each iteration.

venetak replied:
Yes - moved.

alien1976 reviewed (changes requested):
![image](https://user-images.githubusercontent.com/29014573/150354269-1341f5dc-1034-4748-9bd7-d9cbd6b065f9.png)

In cohtml demo page is not working

alien1976 commented on components/tooltip/script.js:
```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
+    }
+
+    overflows() {
+        var rect = this.getBoundingClientRect();
```
```suggestion
        const rect = this.getBoundingClientRect();
```

venetak commented:
> ![image](https://user-images.githubusercontent.com/29014573/150354269-1341f5dc-1034-4748-9bd7-d9cbd6b065f9.png)
> 
> In cohtml demo page is not working

Fixed by adding some delays until the styles are computed and added an animation effect for the showing to avoid weird blinking effects.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
-    setPosition(orientation = this.position) {
-        const elementSize = this.triggerElement.getBoundingClientRect();
+    /**
+     * Calculates the new possition of the tooltip based on the given orientation.
```
```suggestion
     * Calculates the new position of the tooltip based on the given orientation.

```

venetak replied:
Fixed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
-        overflows = overflows || this.overflows();
+        await this.waitForFrames(2);
+        const overflows = this.overflows;
         let overflowingSides = Object.keys(overflows);
```
Why not do `Object.keys(this.overflows)` and remove the declaration from the previous row?

venetak replied:
Yes, removed.

MihailTodorov commented on components/tooltip/script.js:
```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
+     */
+    getVisiblePosition(overflowingSides) {
+        this.uncheckedOrientations = this.uncheckedOrientations.filter(side => !overflowingSides.includes(side));
+        let position = this.uncheckedOrientations[0];
```
```suggestion
        const position = this.uncheckedOrientations.shift();

```
And you can remove the next line

venetak replied:
Nice - I didn't know shift returned the removed element. I also removed the variable, we now directly return `this.uncheckedOrientations.shift();`. Also added a comment just in case to avoid confusion.

alien1976 commented on components/tooltip/script.js:
```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
+        while(count--) {
+            await new Promise((resolve) => requestAnimationFrame(resolve));
+        }
+        return Promise.resolve();
```
You can skip and remove `return Promise.resolve()` because it is async method and it will be resolved automatically.

venetak replied:
Mhmm, removed in both usages.

alien1976 commented on components/tooltip/script.js:
```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
+     */
+    getVisiblePosition(overflowingSides) {
+        this.uncheckedOrientations = this.uncheckedOrientations.filter(side => !overflowingSides.includes(side));
+        let position = this.uncheckedOrientations[0];
+        this.uncheckedOrientations.shift();
+
+        return position;
```
`return this.uncheckedOrientation.shift();` will return the first element and will shift the array. 
[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/shift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/shift)

venetak replied:
Yes, this is the desired behavior. We want to modify the original array. Changed the code a little as suggested by @MihailTodorov. We now directly `return this.uncheckedOrientation.shift()` and the function is shorter, if this is what you meant.
